compiler/rustc_mir/src/transform/promote_consts.rs

```diff
@@ -1268,26 +1268,49 @@ pub fn promote_candidates<'tcx>(
     promotions
 }
 
-/// This function returns `true` if the `const_in_array_repeat_expressions` feature attribute should
-/// be suggested. This function is probably quite expensive, it shouldn't be run in the happy path.
-/// Feature attribute should be suggested if `operand` can be promoted and the feature is not
-/// enabled.
-crate fn should_suggest_const_in_array_repeat_expressions_attribute<'tcx>(
+/// This function returns `true` if the function being called in the array
+/// repeat expression is a `const` function.
+crate fn is_const_fn_in_array_repeat_expression<'tcx>(
     ccx: &ConstCx<'_, 'tcx>,
-    operand: &Operand<'tcx>,
+    place: &Place<'tcx>,
+    body: &Body<'tcx>,
 ) -> bool {
-    let mut rpo = traversal::reverse_postorder(&ccx.body);
-    let (temps, _) = collect_temps_and_candidates(&ccx, &mut rpo);
-    let validator = Validator { ccx, temps: &temps, explicit: false };
-
-    let should_promote = validator.validate_operand(operand).is_ok();
-    let feature_flag = validator.ccx.tcx.features().const_in_array_repeat_expressions;
-    debug!(
-        "should_suggest_const_in_array_repeat_expressions_flag: def_id={:?} \
-            should_promote={:?} feature_flag={:?}",
-        validator.ccx.def_id(),
-        should_promote,
-        feature_flag
-    );
-    should_promote && !feature_flag
+    match place.as_local() {
+        Some(local) => {
+            if body.local_decls[local].is_user_variable() == true {
+                // rule out cases such as: `let my_var = some_fn(); [my_var; N]`
+                return false;
+            }
+        }
+        None => {
+            return false;
+        }
+    }
+
+    for block in body.basic_blocks() {
+        if let Some(Terminator { kind: TerminatorKind::Call { func, destination, .. }, .. }) =
+            &block.terminator
+        {
+            if let Operand::Constant(box Constant { literal: ty::Const { ty, .. }, .. }) = func {
+                if let ty::FnDef(def_id, _) = *ty.kind() {
+                    match destination {
+                        Some((destination_place, _)) => {
+                            if destination_place == place {
+                                if is_const_fn(ccx.tcx, def_id) {
+                                    return true;
+                                } else {
+                                    continue;
```

Comment:
You don't need any of these explicit `continue` expressions, since you don't need to skip anything in the loop body.

Reply:
Done